src/history/hs_rec.c

```diff
@@ -145,40 +135,48 @@ __hs_insert_record(WT_SESSION_IMPL *session, WT_CURSOR *cursor, WT_BTREE *btree,
-                  tw->start_ts == hs_cbt->upd_value->tw.start_ts && tw->start_ts != tw->stop_ts) {
-                    /*
-                     * If we have issues with duplicate history store records, we want to be able to
-                     * distinguish between modifies and full updates. Since modifies are not
-                     * idempotent, having them inserted multiple times can cause invalid values to
-                     * be read.
-                     */
-                    WT_ASSERT(session,
-                      type != WT_UPDATE_MODIFY && (uint8_t)upd_type_full_diag != WT_UPDATE_MODIFY);
+        if (EXTRA_DIAGNOSTICS_ENABLED(session, WT_DIAG_OUT_OF_ORDER | WT_DIAG_VISIBILITY)) {
+            /* Allocate buffer for the existing history store value for the same key. */
+            WT_ERR(__wt_scr_alloc(session, 0, &existing_val));
+
+            if (tw->start_ts == hs_start_ts) {
+                WT_ERR(cursor->get_value(cursor, &hs_stop_durable_ts_diag, &durable_timestamp_diag,
+                  &upd_type_full_diag, existing_val));
+                WT_ERR(__wt_compare(session, NULL, existing_val, hs_value, &cmp));
+                /*
+                 * The same value should not be inserted again unless:
+                 * 1. The previous entry is already deleted (i.e. the stop timestamp is globally
+                 * visible)
+                 * 2. It came from a different transaction
+                 * 3. It came from the same transaction but with a different timestamp
+                 * 4. The prepared rollback left the history store entry when checkpoint is in
+                 * progress.
+                 */
+                if (cmp == 0) {
+                    if (!__wt_txn_tw_stop_visible_all(session, &hs_cbt->upd_value->tw) &&
+                      tw->start_txn != WT_TXN_NONE &&
+                      tw->start_txn == hs_cbt->upd_value->tw.start_txn &&
+                      tw->start_ts == hs_cbt->upd_value->tw.start_ts &&
+                      tw->start_ts != tw->stop_ts) {
+                        /*
+                         * If we have issues with duplicate history store records, we want to be
+                         * able to distinguish between modifies and full updates. Since modifies are
+                         * not idempotent, having them inserted multiple times can cause invalid
+                         * values to be read.
+                         */
+                        WT_ASSERT_ALWAYS(session,
+                          type != WT_UPDATE_MODIFY &&
+                            (uint8_t)upd_type_full_diag != WT_UPDATE_MODIFY,
+                          "Duplicate modifies inserted into the history store can result in "
+                          "invalid reads");
+                    }
                 }
+                counter = hs_counter + 1;
             }
-            counter = hs_counter + 1;
+        } else {
```

Comment:
Is there a reason we left it as an if() inside and else() and didn't make this an else if (tw->start_ts == hs_start_ts)?

Reply:
It was a minor cosmetic request in a previous PR: https://github.com/wiredtiger/wiredtiger/pull/8606#discussion_r1050263518. I'm happy to change it though as it's not a big thing.

Reply:
I remember now, thanks. Happy to leave it as is, was just curious.